Core/Src/main.c

```diff
@@ -1,181 +1,245 @@
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+
+
+  /* USER CODE BEGIN 2 */
+
+  //Start the PWM timer
+  HAL_TIM_Base_Start(&htim1);
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+  //Constants for Clock and ADC output
+  const int CLK_COUNT_PERIOD = 60000;
+  const int MAX_ADC_VAL = 1023;
+  const int MIN_DUTY_CYC = 3000;
+  const int DUTY_CYC_RANGE = 3000;
+  const int TIMEOUT = 100;
+
+  const int EIGHT_BIT_FRAMES = 3;
+
+  //Select the channel and single/differential mode
+  //Send a 00000001 10000000 00000000 binary (CH0) -> 0x01 0x80 0x00 hex
+  const unit8_t TX_DATA [EIGHT_BIT_FRAMES] = {0x01,0x80,0x00};
+  unit8_t rx_data[EIGHT_BIT_FRAMES];
+
+  //Start the Chip select high
+  HAL_GPIO_WritePin (GPIOPB,GPIO_Pin8,1);
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+
+    /* USER CODE BEGIN 3 */
+
+		//Bring the Chip select down to low (start)
+		HAL_GPIO_WritePin (GPIOPB,GPIO_Pin8,0);
+
+		//Transmit receive function to send the transmit selection signal and receive the ADC output from POT
+		HAL_SPI_TransmitReceive(&hspi1, TX_DATA, rx_data, sizeof(EIGHT_BIT_FRAMES), TIMEOUT);
+
+		//Theoretically we have a ADC value now in rx_data [last 10 indexes] 2 from index 1 and 8 from index 2
+			//MSB
+			//theoretical max is 1111111111 = 1023
+			//Start tim1 only once already started
+			//ADC to counts
+			//5-10% duty cycle -> 3000-6000 -> 0-0.33v or 0-330mv
+			//So 1023 = 6000 and 0 = 3000
+
+		//Bring CS up to high after transmit and receive (end)
+		HAL_GPIO_TogglePin (GPIOPB,GPIO_Pin8);
+
+
+		//Use bit shift to get the 10 bit adc value and convert to integer
+		int adc_value = (int)(rx_data [1] << 8 | rx_data[2]);
+
+		//Get the pwm on count using adc ratio and the range of duty cycle
+		unit16_t pwm_count = (unit16_t)(( DUTY_CYC_RANGE * (adc_value / MAX_ADC_VAL))+ MIN_DUTY_CYC);

```

Comment:
the `adc_value` which is type int divided by `MAX_ADC_VAL` here. The result will always be 0 or 1, depending on the compiler, since the two operands are both not floating point value. But the equation looks really good!

Reply:
I added the float conversion.